Stratis.Bitcoin/Builder/BaseFeature.cs

```diff
@@ -0,0 +1,223 @@
+			this._dateTimeProvider = Guard.NotNull(dateTimeProvider, nameof(dateTimeProvider));
+			this._chain = Guard.NotNull(chain, nameof(chain));
+			this._chainBehaviorState = Guard.NotNull(chainState, nameof(chainState));
+			this._signals = Guard.NotNull(signals, nameof(signals));
+			this._connectionManager = Guard.NotNull(connectionManager, nameof(connectionManager));
+		}
+
+		public override void Start()
+		{
+			StartConnectionManager();
+
+			StartAddressManager();
+			StartChain();
+		}
+
+
+		private void StartConnectionManager()
+		{
+			var connectionParameters = _connectionManager.Parameters;
+			connectionParameters.IsRelay = _nodeArgs.Mempool.RelayTxes;
+			connectionParameters.Services = (_nodeArgs.Store.Prune ? NodeServices.Nothing : NodeServices.Network) | NodeServices.NODE_WITNESS;
+
+			_connectionManager = AutoDispose(new ConnectionManager(_network, connectionParameters, _nodeArgs));
+		}
+
+		private void StartChain()
+		{
+			if (!Directory.Exists(_dataFolder.ChainPath))
+			{
+				Logs.FullNode.LogInformation("Creating " + _dataFolder.ChainPath);
+				Directory.CreateDirectory(_dataFolder.ChainPath);
+			}
+
+			var chainRepository = AutoDispose(new ChainRepository(_dataFolder.ChainPath));
+
+			Logs.FullNode.LogInformation("Loading chain");
+
+			chainRepository.Load(_chain).GetAwaiter().GetResult();
+
+			Guard.Assert(_chain.Genesis.HashBlock == _network.GenesisHash); // can't swap networks
+			Logs.FullNode.LogInformation("Chain loaded at height " + _chain.Height);
+
+			_flushChainTask = new PeriodicTask("FlushChain", (cancellation) =>
+			{
+				chainRepository.Save(_chain);
+			})
+			.Start(_cancellationProvider.Cancellation.Token, TimeSpan.FromMinutes(5.0), true);
+
+			AddNodeBehavior(new BlockStore.ChainBehavior(_chain, _chainBehaviorState));
+		}
+
+		private void StartAddressManager()
+		{
+			AddressManager addressManager;
+			if (!File.Exists(_dataFolder.AddrManFile))
+			{
+				Logs.FullNode.LogInformation($"Creating {_dataFolder.AddrManFile}");
+				addressManager = new AddressManager();
+				addressManager.SavePeerFile(_dataFolder.AddrManFile, _network);
+				Logs.FullNode.LogInformation("Created");
+			}
+			else
+			{
+				Logs.FullNode.LogInformation("Loading  {dataFolder.AddrManFile}");
+				addressManager = AddressManager.LoadPeerFile(_dataFolder.AddrManFile);
+				Logs.FullNode.LogInformation("Loaded");
+			}
+
+			if (addressManager.Count == 0)
+			{
+				Logs.FullNode.LogInformation("AddressManager is empty, discovering peers...");
+			}
+
+			_flushAddressManagerTask = new PeriodicTask("FlushAddressManager", (cancellation) =>
+			{
+				addressManager.SavePeerFile(_dataFolder.AddrManFile, _network);
+			})
+		   .Start(_cancellationProvider.Cancellation.Token, TimeSpan.FromMinutes(5.0), true);
+
+
+			AddNodeBehavior(new AddressManagerBehavior(addressManager));
+		}
+
+
+
+		public override void Stop()
+		{
+			_cancellationProvider.Cancellation.Cancel();
```

Comment:
The feature should no cancel the global token source (it should already be cancelled)